{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set dir to /home/dodo/projects/aDNA_Comparative_Analysis\n",
    "import os\n",
    "import pandas as pd\n",
    "os.chdir('../../aDNA_Comparative_Analysis')\n",
    "import re\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.data_loading import load_data, load_mt_dataset\n",
    "from src.data_processing import DataProcessor\n",
    "from src.parsing import match_reich_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# metadata_file -> AmtDB csv metadata file\n",
    "# fasta_amtdb -> AmtDB fasta file\n",
    "# anno_file -> Reich annotation file = metadata file\n",
    "# fasta_reich -> Reich fasta file\n",
    "\n",
    "metadata_file = 'data/amtDB/amtdb_metadata.csv'\n",
    "fasta_amtdb = 'data/amtDB/amtdb_1621-samples_7f_a0pkh.fasta'\n",
    "anno_file = 'data/mitogenomes_reich/v54.1.p1_1240K_public/v54.1.p1_1240K_public.anno'\n",
    "fasta_reich = 'data/mitogenomes_reich/mtdna_reich.fasta'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading AmtDB metadata from 'data/amtDB/amtdb_metadata.csv' and sequence IDs from 'data/amtDB/amtdb_1621-samples_7f_a0pkh.fasta'...\n",
      "Loaded AmtDB metadata with 2541 records and 1621 sequences.\n",
      "\n",
      "Loading 'Reich mt dataset' from 'data/mitogenomes_reich/mtdna_reich.fasta' and metadata from 'data/mitogenomes_reich/v54.1.p1_1240K_public/v54.1.p1_1240K_public.anno'...\n",
      "Loaded 'Reich mt dataset' with 4122 sequences and metadata with 16388 records.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "meta_amtDB, ids_seq_fasta = load_data(metadata_file, fasta_amtdb)\n",
    "ids_mt_dataset, meta_mt_dataset = load_mt_dataset(fasta_reich, anno_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 920 missing sequences.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Process AmtDB data to find missing sequences\n",
    "processor_amtDB = DataProcessor(meta_amtDB, fasta_amtdb)\n",
    "missing_ids_amtDB = processor_amtDB.find_missing_sequences()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "529"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# intersection of missing sequences in AmtDB and Master ID from meta_mt_dataset\n",
    "missing_ids_amtDB_reich = list(set(missing_ids_amtDB) & set(meta_mt_dataset['Master ID']) )\n",
    "len(missing_ids_amtDB_reich)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting sequences matching the specified IDs from 'data/mitogenomes_reich/mtdna_reich.fasta'...\n",
      "Saved 404 sequences to 'output/missing_sequences_AmtDB.fasta'.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# Extract missing sequences from the Reich dataset (fasta_reich) and save them\n",
    "output_fasta = \"output/missing_sequences_AmtDB.fasta\"\n",
    "processor_amtDB.extract_and_save_sequences(fasta_reich, missing_ids_amtDB, output_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Haplogrep 3 3.2.1\n",
      "(c) 2022-2023 Sebastian Sch√∂nherr, Hansi Weissensteiner, Lukas Forer\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "Written haplogroups to file output/analysis_result.hsd\n"
     ]
    }
   ],
   "source": [
    "!haplogrep3 classify --tree phylotree-rcrs@17.0 --in output/missing_sequences_AmtDB.fasta --out output/analysis_result.hsd --extend-report\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Haplogroup</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Range</th>\n",
       "      <th>Not_Found_Polys</th>\n",
       "      <th>Found_Polys</th>\n",
       "      <th>Remaining_Polys</th>\n",
       "      <th>AAC_In_Remainings</th>\n",
       "      <th>Input_Sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I0070</td>\n",
       "      <td>H13a1a</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9691</td>\n",
       "      <td>1-16569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>263G 750G 1438G 2259T 4745G 4769G 8860G 13680T...</td>\n",
       "      <td>310C (localPrivateMutation) 3107T (globalPriva...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>263G 310C 750G 1438G 2259T 3107T 4745G 4769G 8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I0071</td>\n",
       "      <td>U5a1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.9644</td>\n",
       "      <td>1-16569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73G 263G 750G 1438G 2706G 3197C 4769G 7028T 88...</td>\n",
       "      <td>310C (localPrivateMutation) 3107T (globalPriva...</td>\n",
       "      <td>8705C [M60T| Codon 2 | ATP6 ]</td>\n",
       "      <td>73G 263G 310C 750G 1438G 2706G 3107T 3197C 476...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SampleID Haplogroup  Rank  Quality    Range Not_Found_Polys  \\\n",
       "0    I0070     H13a1a     1   0.9691  1-16569             NaN   \n",
       "1    I0071       U5a1     1   0.9644  1-16569             NaN   \n",
       "\n",
       "                                         Found_Polys  \\\n",
       "0  263G 750G 1438G 2259T 4745G 4769G 8860G 13680T...   \n",
       "1  73G 263G 750G 1438G 2706G 3197C 4769G 7028T 88...   \n",
       "\n",
       "                                     Remaining_Polys  \\\n",
       "0  310C (localPrivateMutation) 3107T (globalPriva...   \n",
       "1  310C (localPrivateMutation) 3107T (globalPriva...   \n",
       "\n",
       "               AAC_In_Remainings  \\\n",
       "0                            NaN   \n",
       "1  8705C [M60T| Codon 2 | ATP6 ]   \n",
       "\n",
       "                                        Input_Sample  \n",
       "0  263G 310C 750G 1438G 2259T 3107T 4745G 4769G 8...  \n",
       "1  73G 263G 310C 750G 1438G 2706G 3107T 3197C 476...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load analysis_result.hsd\n",
    "analysis_result = pd.read_csv('output/analysis_result.hsd', sep='\\t')\n",
    "analysis_result.head(2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['SampleID', 'Haplogroup', 'Rank', 'Quality', 'Range', 'Not_Found_Polys',\n",
       "       'Found_Polys', 'Remaining_Polys', 'AAC_In_Remainings', 'Input_Sample'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# columns\n",
    "analysis_result.columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Rename 'SampleID' column to 'ID'\n",
    "analysis_result.rename(columns={'SampleID': 'ID'}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use 'Found_Polys' as the 'Polymorphisms' column directly\n",
    "# Assuming 'Found_Polys' accurately represents the polymorphisms we want to include\n",
    "# If a combination of columns is needed, you may need to concatenate them appropriately here\n",
    "analysis_result['Polymorphisms'] = analysis_result['Found_Polys']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "###########################################################################\n",
      "# MitoPathoPy - Tool for annotating pathological mutations in human mtDNA #\n",
      "###########################################################################\n",
      "Author: Edvard Ehler, PhD (edvard.ehler@img.cas.cz)\n",
      "Year: 2020\n",
      "Source: https://github.com/EdaEhler/MitoPathoPy\n",
      "\n",
      "This tool will search for the (potential) pathological mutations in you mtDNA samples. To get the right input format, please, use the Haplogrep 2 (https://haplogrep.i-med.ac.at/app/index.html) tool to turn your sequences (fastas) into HSD format file. Haplogrep can be also downloaded and run localy (https://github.com/seppinho/haplogrep-cmd). Pathological mtDNA mutations are taken from data published at https://www.mitomap.org/MITOMAP.\n",
      "\n",
      "We invite you to check our ancient mtDNA database at: https://amtdb.org/\n",
      "\n",
      "If you use this tool in you research, please, consider citing our article:\n",
      "Ehler E, Novotn√Ω J, Juras A, Chyle≈Ñski M, Moravƒç√≠k O, Paƒçes J. AmtDB: a database of ancient human mitochondrial genomes. Nucleic acids research. 2019 Jan 8;47(D1):D29-32.\n",
      "\n",
      "###########################################################################\n",
      "    \n"
     ]
    }
   ],
   "source": [
    "!python3 mitopatho/mitopatho.py -i output/processed_haplogroups.hsd -o output/mitopatho_output.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load mitopatho_output.txt\n",
    "mitopatho_output = pd.read_csv('output/mitopatho_output.txt', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'11467G,Altered brain pH / sCJD patients,Reported;12308G,CPEO / Stroke / CM / Breast & Renal & Prostate Cancer Risk / Altered brain pH /sCJD,Reported;12372A,Altered brain pH / sCJD patients,Reported;16192T,Melanoma patients,Reported;16270T,Melanoma patients,Reported'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pprint first two rows of the DataFrame but full data \n",
    "id_1 = mitopatho_output[mitopatho_output['ID'] == 'I0071']\n",
    "# print second column of the first row\n",
    "id_1.iloc[0, 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 920 missing sequences.\n",
      "\n",
      "Loading 'Reich mt dataset' from 'data/mitogenomes_reich/mtdna_reich.fasta' and metadata from 'data/mitogenomes_reich/v54.1.p1_1240K_public/v54.1.p1_1240K_public.anno'...\n",
      "Loaded 'Reich mt dataset' with 4122 sequences and metadata with 16388 records.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "missing_ids_amtDB = processor_amtDB.find_missing_sequences()\n",
    "ids_mt_dataset, meta_mt_dataset = load_mt_dataset(fasta_reich, anno_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of missing sequences in AmtDB that are in the Reich dataset: 529\n"
     ]
    }
   ],
   "source": [
    "# how many missing_ids_amtDB in the reich dataset\n",
    "missing_ids_amtDB_in_reich = [id for id in missing_ids_amtDB if id in meta_mt_dataset['Master ID'].values]\n",
    "print(f\"Number of missing sequences in AmtDB that are in the Reich dataset: {len(missing_ids_amtDB_in_reich)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Polymorphism</th>\n",
       "      <th>Position</th>\n",
       "      <th>Status</th>\n",
       "      <th>Locus</th>\n",
       "      <th>Diseases</th>\n",
       "      <th>Homoplasmy</th>\n",
       "      <th>Heteroplasmy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I0071</td>\n",
       "      <td>11467G</td>\n",
       "      <td>11467</td>\n",
       "      <td>Reported</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>Altered brain pH / sCJD patients</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I0071</td>\n",
       "      <td>12308G</td>\n",
       "      <td>12308</td>\n",
       "      <td>Reported</td>\n",
       "      <td>MT-TL2</td>\n",
       "      <td>CPEO / Stroke / CM / Breast &amp; Renal &amp; Prostate...</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I0071</td>\n",
       "      <td>12372A</td>\n",
       "      <td>12372</td>\n",
       "      <td>Reported</td>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>Altered brain pH / sCJD patients</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I0071</td>\n",
       "      <td>16192T</td>\n",
       "      <td>16192</td>\n",
       "      <td>Reported</td>\n",
       "      <td>MT-CR</td>\n",
       "      <td>Melanoma patients</td>\n",
       "      <td>nr</td>\n",
       "      <td>nr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I0071</td>\n",
       "      <td>16270T</td>\n",
       "      <td>16270</td>\n",
       "      <td>Reported</td>\n",
       "      <td>MT-CR</td>\n",
       "      <td>Melanoma patients</td>\n",
       "      <td>nr</td>\n",
       "      <td>nr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1704</th>\n",
       "      <td>I9131</td>\n",
       "      <td>10398G</td>\n",
       "      <td>10398</td>\n",
       "      <td>Reported / lineage L &amp; M marker / also hg IJK</td>\n",
       "      <td>MT-ND3</td>\n",
       "      <td>PD protective factor / longevity / altered cel...</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1705</th>\n",
       "      <td>I9131</td>\n",
       "      <td>11467G</td>\n",
       "      <td>11467</td>\n",
       "      <td>Reported</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>Altered brain pH / sCJD patients</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1706</th>\n",
       "      <td>I9131</td>\n",
       "      <td>12308G</td>\n",
       "      <td>12308</td>\n",
       "      <td>Reported</td>\n",
       "      <td>MT-TL2</td>\n",
       "      <td>CPEO / Stroke / CM / Breast &amp; Renal &amp; Prostate...</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1707</th>\n",
       "      <td>I9131</td>\n",
       "      <td>12372A</td>\n",
       "      <td>12372</td>\n",
       "      <td>Reported</td>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>Altered brain pH / sCJD patients</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1708</th>\n",
       "      <td>I9131</td>\n",
       "      <td>9055A</td>\n",
       "      <td>9055</td>\n",
       "      <td>Reported</td>\n",
       "      <td>MT-ATP6</td>\n",
       "      <td>PD protective factor</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1709 rows √ó 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID Polymorphism  Position  \\\n",
       "0     I0071       11467G     11467   \n",
       "1     I0071       12308G     12308   \n",
       "2     I0071       12372A     12372   \n",
       "3     I0071       16192T     16192   \n",
       "4     I0071       16270T     16270   \n",
       "...     ...          ...       ...   \n",
       "1704  I9131       10398G     10398   \n",
       "1705  I9131       11467G     11467   \n",
       "1706  I9131       12308G     12308   \n",
       "1707  I9131       12372A     12372   \n",
       "1708  I9131        9055A      9055   \n",
       "\n",
       "                                             Status    Locus  \\\n",
       "0                                          Reported   MT-ND4   \n",
       "1                                          Reported   MT-TL2   \n",
       "2                                          Reported   MT-ND5   \n",
       "3                                          Reported    MT-CR   \n",
       "4                                          Reported    MT-CR   \n",
       "...                                             ...      ...   \n",
       "1704  Reported / lineage L & M marker / also hg IJK   MT-ND3   \n",
       "1705                                       Reported   MT-ND4   \n",
       "1706                                       Reported   MT-TL2   \n",
       "1707                                       Reported   MT-ND5   \n",
       "1708                                       Reported  MT-ATP6   \n",
       "\n",
       "                                               Diseases Homoplasmy  \\\n",
       "0                      Altered brain pH / sCJD patients          +   \n",
       "1     CPEO / Stroke / CM / Breast & Renal & Prostate...          +   \n",
       "2                      Altered brain pH / sCJD patients          +   \n",
       "3                                     Melanoma patients         nr   \n",
       "4                                     Melanoma patients         nr   \n",
       "...                                                 ...        ...   \n",
       "1704  PD protective factor / longevity / altered cel...          +   \n",
       "1705                   Altered brain pH / sCJD patients          +   \n",
       "1706  CPEO / Stroke / CM / Breast & Renal & Prostate...          +   \n",
       "1707                   Altered brain pH / sCJD patients          +   \n",
       "1708                               PD protective factor          +   \n",
       "\n",
       "     Heteroplasmy  \n",
       "0               -  \n",
       "1               +  \n",
       "2               -  \n",
       "3              nr  \n",
       "4              nr  \n",
       "...           ...  \n",
       "1704            -  \n",
       "1705            -  \n",
       "1706            +  \n",
       "1707            -  \n",
       "1708            -  \n",
       "\n",
       "[1709 rows x 8 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mitopatho_csv_path = 'output/AmtDB MitoPathoTool.csv'\n",
    "mitop_data = pd.read_csv(mitopatho_csv_path)\n",
    "mitop_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sequence IDs matched from mitop_data and missing_ids_amtDB_in_reich: 325\n"
     ]
    }
   ],
   "source": [
    "# number of sequence IDs matched from mitop_data and missing_ids_amtDB_in_reich\n",
    "matched_ids = [id for id in missing_ids_amtDB_in_reich if id in mitop_data['ID'].values]\n",
    "print(f\"Number of sequence IDs matched from mitop_data and missing_ids_amtDB_in_reich: {len(matched_ids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading Reich metadata...\n",
      "Reich metadata has 16388 rows.\n",
      "MitoPatho data has 1709 rows.\n",
      "Processed 561 out of 920 missing IDs.\n",
      "Total matched rows: 561\n"
     ]
    }
   ],
   "source": [
    "# Example of how to call the function with paths to your CSV files\n",
    "reich_meta_file_path = 'data/mitogenomes_reich/v54.1.p1_1240K_public/v54.1.p1_1240K_public.anno'\n",
    "mitopatho_csv_path = 'output/AmtDB MitoPathoTool.csv'\n",
    "missing_ids = missing_ids_amtDB\n",
    "\n",
    "# Now, match metadata with the updated function that includes MitoPatho data\n",
    "matched_metadata_df = match_reich_metadata(reich_meta_file_path, missing_ids, mitopatho_csv_path)\n",
    "print(f\"Total matched rows: {len(matched_metadata_df)}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>alternative_identifiers</th>\n",
       "      <th>country</th>\n",
       "      <th>continent</th>\n",
       "      <th>region</th>\n",
       "      <th>culture</th>\n",
       "      <th>epoch</th>\n",
       "      <th>group</th>\n",
       "      <th>comment</th>\n",
       "      <th>latitude</th>\n",
       "      <th>...</th>\n",
       "      <th>ychr_snps</th>\n",
       "      <th>avg_coverage</th>\n",
       "      <th>sequence_source</th>\n",
       "      <th>mitopatho_alleles</th>\n",
       "      <th>mitopatho_positions</th>\n",
       "      <th>mitopatho_locus</th>\n",
       "      <th>mitopatho_diseases</th>\n",
       "      <th>mitopatho_statuses</th>\n",
       "      <th>mitopatho_homoplasms</th>\n",
       "      <th>mitopatho_heteroplasms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I1072</td>\n",
       "      <td>I1072_enhanced_d</td>\n",
       "      <td>Israel</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Israel_Natufian_d</td>\n",
       "      <td></td>\n",
       "      <td>Israel_Natufian_d</td>\n",
       "      <td></td>\n",
       "      <td>32.65</td>\n",
       "      <td>...</td>\n",
       "      <td>..</td>\n",
       "      <td>8.193132</td>\n",
       "      <td>fasta</td>\n",
       "      <td>12705T;16390A;8836G</td>\n",
       "      <td>12705;16390;8836</td>\n",
       "      <td>MT-CO1;MT-CR;MT-ATP6</td>\n",
       "      <td>Possible protective factor for normal tension ...</td>\n",
       "      <td>Reported</td>\n",
       "      <td>NaN;+</td>\n",
       "      <td>NaN;-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I1072</td>\n",
       "      <td>I1072_d</td>\n",
       "      <td>Israel</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Israel_Natufian</td>\n",
       "      <td></td>\n",
       "      <td>Israel_Natufian</td>\n",
       "      <td></td>\n",
       "      <td>32.65</td>\n",
       "      <td>...</td>\n",
       "      <td>..</td>\n",
       "      <td>1.897519</td>\n",
       "      <td>fasta</td>\n",
       "      <td>12705T;16390A;8836G</td>\n",
       "      <td>12705;16390;8836</td>\n",
       "      <td>MT-CO1;MT-CR;MT-ATP6</td>\n",
       "      <td>Possible protective factor for normal tension ...</td>\n",
       "      <td>Reported</td>\n",
       "      <td>NaN;+</td>\n",
       "      <td>NaN;-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I1072</td>\n",
       "      <td>I1072</td>\n",
       "      <td>Israel</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Israel_Natufian_contam</td>\n",
       "      <td></td>\n",
       "      <td>Israel_Natufian_contam</td>\n",
       "      <td></td>\n",
       "      <td>32.65</td>\n",
       "      <td>...</td>\n",
       "      <td>..</td>\n",
       "      <td>12.656105</td>\n",
       "      <td>fasta</td>\n",
       "      <td>12705T;16390A;8836G</td>\n",
       "      <td>12705;16390;8836</td>\n",
       "      <td>MT-CO1;MT-CR;MT-ATP6</td>\n",
       "      <td>Possible protective factor for normal tension ...</td>\n",
       "      <td>Reported</td>\n",
       "      <td>NaN;+</td>\n",
       "      <td>NaN;-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I1415</td>\n",
       "      <td>I1415_enhanced</td>\n",
       "      <td>Jordan</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Jordan_PPNB</td>\n",
       "      <td></td>\n",
       "      <td>Jordan_PPNB</td>\n",
       "      <td></td>\n",
       "      <td>31.988</td>\n",
       "      <td>...</td>\n",
       "      <td>..</td>\n",
       "      <td>23.444746</td>\n",
       "      <td>fasta</td>\n",
       "      <td>11251G;15928A;16189C;4216C;4917G</td>\n",
       "      <td>11251;15928;16189;4216;4917</td>\n",
       "      <td>MT-ND4;MT-TT;MT-CR;MT-ND1;MT-ND2</td>\n",
       "      <td>Reduced risk of PD;Multiple Sclerosis / idiopa...</td>\n",
       "      <td>Reported;Conflicting reports</td>\n",
       "      <td>nr;+</td>\n",
       "      <td>nr;-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I1701</td>\n",
       "      <td>I1701_enhanced</td>\n",
       "      <td>Jordan</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Jordan_PPNB</td>\n",
       "      <td></td>\n",
       "      <td>Jordan_PPNB</td>\n",
       "      <td></td>\n",
       "      <td>31.988</td>\n",
       "      <td>...</td>\n",
       "      <td>n/a (female)</td>\n",
       "      <td>14.818215</td>\n",
       "      <td>fasta</td>\n",
       "      <td>10398G;11467G;12308G;12372A;16093C;9055A</td>\n",
       "      <td>10398;11467;12308;12372;16093;9055</td>\n",
       "      <td>MT-ND3;MT-ND4;MT-TL2;MT-ND5;MT-CR;MT-ATP6</td>\n",
       "      <td>PD protective factor / longevity / altered cel...</td>\n",
       "      <td>Reported / lineage L &amp; M marker / also hg IJK;...</td>\n",
       "      <td>+;-</td>\n",
       "      <td>-;+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>556</th>\n",
       "      <td>ILK002</td>\n",
       "      <td>ILK002</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Ukraine_EBA_GlobularAmphora</td>\n",
       "      <td></td>\n",
       "      <td>Ukraine_EBA_GlobularAmphora</td>\n",
       "      <td></td>\n",
       "      <td>49.55777778</td>\n",
       "      <td>...</td>\n",
       "      <td>..</td>\n",
       "      <td>..</td>\n",
       "      <td>fasta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>557</th>\n",
       "      <td>ILK001</td>\n",
       "      <td>ILK001</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Ukraine_GlobularAmphora</td>\n",
       "      <td></td>\n",
       "      <td>Ukraine_GlobularAmphora</td>\n",
       "      <td></td>\n",
       "      <td>49.55777778</td>\n",
       "      <td>...</td>\n",
       "      <td>..</td>\n",
       "      <td>..</td>\n",
       "      <td>fasta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558</th>\n",
       "      <td>ILK003</td>\n",
       "      <td>ILK003</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Ukraine_GlobularAmphora</td>\n",
       "      <td></td>\n",
       "      <td>Ukraine_GlobularAmphora</td>\n",
       "      <td></td>\n",
       "      <td>49.55777778</td>\n",
       "      <td>...</td>\n",
       "      <td>n/a (female)</td>\n",
       "      <td>..</td>\n",
       "      <td>fasta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>559</th>\n",
       "      <td>I1944</td>\n",
       "      <td>S1944.E1.L3_v54.1_addback</td>\n",
       "      <td>Iran</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Iran_GanjDareh_N</td>\n",
       "      <td></td>\n",
       "      <td>Iran_GanjDareh_N</td>\n",
       "      <td></td>\n",
       "      <td>34.45</td>\n",
       "      <td>...</td>\n",
       "      <td>..</td>\n",
       "      <td>4.453256</td>\n",
       "      <td>fasta</td>\n",
       "      <td>4216C</td>\n",
       "      <td>4216</td>\n",
       "      <td>MT-ND1</td>\n",
       "      <td>LHON / Insulin Resistance /possible adaptive h...</td>\n",
       "      <td>Conflicting reports</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>560</th>\n",
       "      <td>I1951</td>\n",
       "      <td>S1951.E1.L3_v54.1_addback</td>\n",
       "      <td>Iran</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Iran_GanjDareh_N</td>\n",
       "      <td></td>\n",
       "      <td>Iran_GanjDareh_N</td>\n",
       "      <td></td>\n",
       "      <td>34.45</td>\n",
       "      <td>...</td>\n",
       "      <td>..</td>\n",
       "      <td>13.102963</td>\n",
       "      <td>fasta</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>561 rows √ó 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    identifier    alternative_identifiers  country continent region  \\\n",
       "0        I1072           I1072_enhanced_d   Israel                    \n",
       "1        I1072                    I1072_d   Israel                    \n",
       "2        I1072                      I1072   Israel                    \n",
       "3        I1415             I1415_enhanced   Jordan                    \n",
       "4        I1701             I1701_enhanced   Jordan                    \n",
       "..         ...                        ...      ...       ...    ...   \n",
       "556     ILK002                     ILK002  Ukraine                    \n",
       "557     ILK001                     ILK001  Ukraine                    \n",
       "558     ILK003                     ILK003  Ukraine                    \n",
       "559      I1944  S1944.E1.L3_v54.1_addback     Iran                    \n",
       "560      I1951  S1951.E1.L3_v54.1_addback     Iran                    \n",
       "\n",
       "                         culture epoch                        group comment  \\\n",
       "0              Israel_Natufian_d                  Israel_Natufian_d           \n",
       "1                Israel_Natufian                    Israel_Natufian           \n",
       "2         Israel_Natufian_contam             Israel_Natufian_contam           \n",
       "3                    Jordan_PPNB                        Jordan_PPNB           \n",
       "4                    Jordan_PPNB                        Jordan_PPNB           \n",
       "..                           ...   ...                          ...     ...   \n",
       "556  Ukraine_EBA_GlobularAmphora        Ukraine_EBA_GlobularAmphora           \n",
       "557      Ukraine_GlobularAmphora            Ukraine_GlobularAmphora           \n",
       "558      Ukraine_GlobularAmphora            Ukraine_GlobularAmphora           \n",
       "559             Iran_GanjDareh_N                   Iran_GanjDareh_N           \n",
       "560             Iran_GanjDareh_N                   Iran_GanjDareh_N           \n",
       "\n",
       "        latitude  ...     ychr_snps avg_coverage sequence_source  \\\n",
       "0          32.65  ...            ..     8.193132           fasta   \n",
       "1          32.65  ...            ..     1.897519           fasta   \n",
       "2          32.65  ...            ..    12.656105           fasta   \n",
       "3         31.988  ...            ..    23.444746           fasta   \n",
       "4         31.988  ...  n/a (female)    14.818215           fasta   \n",
       "..           ...  ...           ...          ...             ...   \n",
       "556  49.55777778  ...            ..           ..           fasta   \n",
       "557  49.55777778  ...            ..           ..           fasta   \n",
       "558  49.55777778  ...  n/a (female)           ..           fasta   \n",
       "559        34.45  ...            ..     4.453256           fasta   \n",
       "560        34.45  ...            ..    13.102963           fasta   \n",
       "\n",
       "                            mitopatho_alleles  \\\n",
       "0                         12705T;16390A;8836G   \n",
       "1                         12705T;16390A;8836G   \n",
       "2                         12705T;16390A;8836G   \n",
       "3            11251G;15928A;16189C;4216C;4917G   \n",
       "4    10398G;11467G;12308G;12372A;16093C;9055A   \n",
       "..                                        ...   \n",
       "556                                             \n",
       "557                                             \n",
       "558                                             \n",
       "559                                     4216C   \n",
       "560                                             \n",
       "\n",
       "                    mitopatho_positions  \\\n",
       "0                      12705;16390;8836   \n",
       "1                      12705;16390;8836   \n",
       "2                      12705;16390;8836   \n",
       "3           11251;15928;16189;4216;4917   \n",
       "4    10398;11467;12308;12372;16093;9055   \n",
       "..                                  ...   \n",
       "556                                       \n",
       "557                                       \n",
       "558                                       \n",
       "559                                4216   \n",
       "560                                       \n",
       "\n",
       "                               mitopatho_locus  \\\n",
       "0                         MT-CO1;MT-CR;MT-ATP6   \n",
       "1                         MT-CO1;MT-CR;MT-ATP6   \n",
       "2                         MT-CO1;MT-CR;MT-ATP6   \n",
       "3             MT-ND4;MT-TT;MT-CR;MT-ND1;MT-ND2   \n",
       "4    MT-ND3;MT-ND4;MT-TL2;MT-ND5;MT-CR;MT-ATP6   \n",
       "..                                         ...   \n",
       "556                                              \n",
       "557                                              \n",
       "558                                              \n",
       "559                                     MT-ND1   \n",
       "560                                              \n",
       "\n",
       "                                    mitopatho_diseases  \\\n",
       "0    Possible protective factor for normal tension ...   \n",
       "1    Possible protective factor for normal tension ...   \n",
       "2    Possible protective factor for normal tension ...   \n",
       "3    Reduced risk of PD;Multiple Sclerosis / idiopa...   \n",
       "4    PD protective factor / longevity / altered cel...   \n",
       "..                                                 ...   \n",
       "556                                                      \n",
       "557                                                      \n",
       "558                                                      \n",
       "559  LHON / Insulin Resistance /possible adaptive h...   \n",
       "560                                                      \n",
       "\n",
       "                                    mitopatho_statuses mitopatho_homoplasms  \\\n",
       "0                                             Reported                NaN;+   \n",
       "1                                             Reported                NaN;+   \n",
       "2                                             Reported                NaN;+   \n",
       "3                         Reported;Conflicting reports                 nr;+   \n",
       "4    Reported / lineage L & M marker / also hg IJK;...                  +;-   \n",
       "..                                                 ...                  ...   \n",
       "556                                                                           \n",
       "557                                                                           \n",
       "558                                                                           \n",
       "559                                Conflicting reports                    +   \n",
       "560                                                                           \n",
       "\n",
       "    mitopatho_heteroplasms  \n",
       "0                    NaN;-  \n",
       "1                    NaN;-  \n",
       "2                    NaN;-  \n",
       "3                     nr;-  \n",
       "4                      -;+  \n",
       "..                     ...  \n",
       "556                         \n",
       "557                         \n",
       "558                         \n",
       "559                      -  \n",
       "560                         \n",
       "\n",
       "[561 rows x 33 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#  save the matched metadata to a file\n",
    "matched_metadata_df.to_csv('output/matched_metadata.csv', index=False)\n",
    "matched_metadata_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['I4188_enhanced',\n",
       " 'I16170_enhanced',\n",
       " 'I16170',\n",
       " 'I15947',\n",
       " 'I7341',\n",
       " 'I3315_enhanced',\n",
       " 'I13983',\n",
       " 'I13983_enhanced_d',\n",
       " 'I13983',\n",
       " 'I13983_enhanced_d',\n",
       " 'I15947',\n",
       " 'I16170',\n",
       " 'I16170_enhanced',\n",
       " 'I3315_enhanced',\n",
       " 'I4188_enhanced',\n",
       " 'I7341']"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load fasta file missing_sequences_AmtDB.fasta\n",
    "fasta_file = 'output/missing_sequences_AmtDB.fasta'\n",
    "fasta_reich = 'data/mitogenomes_reich/mtdna_reich.fasta'\n",
    "# load fasta file\n",
    "sequences = list(SeqIO.parse(fasta_reich, 'fasta'))\n",
    "id_sequences = {seq.id for seq in sequences}\n",
    "# from id_sequences, find the ids that are duplicated\n",
    "id_sequences = [seq.id for seq in sequences]\n",
    "duplicated_ids = [id for id in id_sequences if id_sequences.count(id) > 1]\n",
    "duplicated_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracting sequences matching the specified IDs from 'data/mitogenomes_reich/mtdna_reich.fasta'...\n",
      "Saved 16 sequences to 'output/duplicates.fasta'.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# extract these sequences from the fasta file and store them in a new file duplicates.fasta\n",
    "output_fasta = \"output/duplicates.fasta\"\n",
    "processor_amtDB.extract_and_save_sequences(fasta_reich, duplicated_ids, output_fasta)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aDNA_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
